packages/docusaurus-mdx-loader/src/remark/toc/index.ts

```diff
@@ -93,21 +93,75 @@ const plugin: Plugin = function plugin(
     const {toString} = await import('mdast-util-to-string');
     const {visit} = await import('unist-util-visit');
 
-    const headings: TOCItem[] = [];
+    const partialComponentToHeadingsName: {[key: string]: string} =
+      Object.create(null);
 
-    visit(root, 'heading', (child: Heading) => {
-      const value = toString(child);
+    // TOCItem or string already with the spread operator
+    const headings: (TOCItem | string)[] = [];
 
-      // depth:1 headings are titles and not included in the TOC
-      if (!value || child.depth < 2) {
-        return;
+    visit(root, ['heading', 'mdxjsEsm', 'mdxJsxFlowElement'], (child) => {
+      if (child.type === 'heading') {
+        const headingNode = child as Heading;
+        const value = toString(headingNode);
+
+        // depth:1 headings are titles and not included in the TOC
+        if (!value || headingNode.depth < 2) {
+          return;
+        }
+
+        headings.push({
+          value: toValue(headingNode, toString),
+          id: headingNode.data!.id!,
+          level: headingNode.depth,
+        });
       }
 
-      headings.push({
-        value: toValue(child, toString),
-        id: child.data!.id!,
-        level: child.depth,
-      });
+      if (child.type === 'mdxjsEsm') {
+        const importNode = child as MdxjsEsm;
+        if (!importNode.data?.estree) {
+          return;
+        }
+
+        for (const importDeclaration of importNode.data.estree.body) {
+          if (importDeclaration.type !== 'ImportDeclaration') {
+            continue;
+          }
+          const importPath = importDeclaration.source.value as string;
+          const isMdxImport = /\.mdx?$/.test(importPath);
+          if (!isMdxImport) {
+            continue;
+          }
+
+          const componentName = importDeclaration.specifiers.find(
+            (o: any) => o.type === 'ImportDefaultSpecifier',
+          )?.local.name;
+
+          if (!componentName) {
+            continue;
+          }
+          const {length} = Object.keys(partialComponentToHeadingsName);
+          const exportAsName = `${name}${length}`;
+          partialComponentToHeadingsName[componentName] = exportAsName;
+
+          importDeclaration.specifiers.push({
+            type: 'ImportSpecifier',
+            imported: {type: 'Identifier', name},
+            local: {type: 'Identifier', name: exportAsName},
+          });
+        }
+      }
+
+      if (child.type === 'mdxJsxFlowElement') {
+        const node = child as MdxJsxFlowElement;
+        const nodeName = node.name;
+        if (!nodeName) {
+          return;
+        }
+        const headingsName = partialComponentToHeadingsName[nodeName];
+        if (headingsName) {
+          headings.push(`...${headingsName}`);
```

Comment:
Not super fan of using this "string-based data structure": you add "..." to later remove this prefix. Maybe use some kind of object instead?

Reply:
Came up with a new type called `NestedTOC`, that is used like this:
```typescript
type NestedTOC = {
  readonly nested: true;
  readonly name: string;
};

const headings: (TOCItem | NestedTOC)[] = [];
``` 

This is because I need to store both types in one array, but at some point I have to process them differently.

Can you give your thoughts on this?

Reply:
I don't really have a better name and it's only a local type so that's good enough to me 👍